---a/net/minecraft/item/crafting/RecipeBookCloning.java
+++b/net/minecraft/item/crafting/RecipeBookCloning.java
@@ -7,31 +7,32 @@
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.world.World;
 
-public class RecipeBookCloning implements IRecipe
-{
+public class RecipeBookCloning extends ShapelessRecipes implements IRecipe { // CraftBukkit - added extends
+
     private static final String __OBFID = "CL_00000081";
 
-    public boolean matches(InventoryCrafting p_77569_1_, World worldIn)
-    {
+    // CraftBukkit start - Delegate to new parent class
+    public RecipeBookCloning() {
+        super(new ItemStack(Items.written_book, 0, -1), java.util.Arrays.asList(new ItemStack(Items.writable_book, 0, 0)));
+    }
+    // CraftBukkit end
+
+    public boolean matches(InventoryCrafting p_77569_1_, World worldIn) {
         int i = 0;
         ItemStack itemstack = null;
 
-        for (int j = 0; j < p_77569_1_.getSizeInventory(); ++j)
-        {
+        for (int j = 0; j < p_77569_1_.getSizeInventory(); ++j) {
             ItemStack itemstack1 = p_77569_1_.getStackInSlot(j);
 
-            if (itemstack1 != null)
-            {
-                if (itemstack1.getItem() == Items.written_book)
-                {
-                    if (itemstack != null)
-                    {
+            if (itemstack1 != null) {
+                if (itemstack1.getItem() == Items.written_book) {
+                    if (itemstack != null) {
                         return false;
                     }
 
                     itemstack = itemstack1;
-                }
-                else
-                {
-                    if (itemstack1.getItem() != Items.writable_book)
-                    {
+                } else {
+                    if (itemstack1.getItem() != Items.writable_book) {
                         return false;
                     }
 
@@ -46,27 +46,21 @@
         return itemstack != null && i > 0;
     }
 
-    public ItemStack getCraftingResult(InventoryCrafting p_77572_1_)
-    {
+    public ItemStack getCraftingResult(InventoryCrafting p_77572_1_) {
         int i = 0;
         ItemStack itemstack = null;
 
-        for (int j = 0; j < p_77572_1_.getSizeInventory(); ++j)
-        {
+        for (int j = 0; j < p_77572_1_.getSizeInventory(); ++j) {
             ItemStack itemstack1 = p_77572_1_.getStackInSlot(j);
 
-            if (itemstack1 != null)
-            {
-                if (itemstack1.getItem() == Items.written_book)
-                {
-                    if (itemstack != null)
-                    {
+            if (itemstack1 != null) {
+                if (itemstack1.getItem() == Items.written_book) {
+                    if (itemstack != null) {
                         return null;
                     }
 
                     itemstack = itemstack1;
-                }
-                else
-                {
-                    if (itemstack1.getItem() != Items.writable_book)
-                    {
+                } else {
+                    if (itemstack1.getItem() != Items.writable_book) {
                         return null;
                     }
 
@@ -78,45 +78,36 @@
             }
         }
 
-        if (itemstack != null && i >= 1 && ItemEditableBook.getGeneration(itemstack) < 2)
-        {
+        if (itemstack != null && i >= 1 && ItemEditableBook.getGeneration(itemstack) < 2) {
             ItemStack itemstack2 = new ItemStack(Items.written_book, i);
-            itemstack2.setTagCompound((NBTTagCompound)itemstack.getTagCompound().copy());
+            itemstack2.setTagCompound((NBTTagCompound) itemstack.getTagCompound().copy());
             itemstack2.getTagCompound().setInteger("generation", ItemEditableBook.getGeneration(itemstack) + 1);
 
-            if (itemstack.hasDisplayName())
-            {
+            if (itemstack.hasDisplayName()) {
                 itemstack2.setStackDisplayName(itemstack.getDisplayName());
             }
 
             return itemstack2;
-        }
-        else
-        {
+        } else {
             return null;
         }
     }
 
-    public int getRecipeSize()
-    {
+    public int getRecipeSize() {
         return 9;
     }
 
-    public ItemStack getRecipeOutput()
-    {
+    public ItemStack getRecipeOutput() {
         return null;
     }
 
-    public ItemStack[] getRemainingItems(InventoryCrafting p_179532_1_)
-    {
+    public ItemStack[] getRemainingItems(InventoryCrafting p_179532_1_) {
         ItemStack[] aitemstack = new ItemStack[p_179532_1_.getSizeInventory()];
 
-        for (int i = 0; i < aitemstack.length; ++i)
-        {
+        for (int i = 0; i < aitemstack.length; ++i) {
             ItemStack itemstack = p_179532_1_.getStackInSlot(i);
 
-            if (itemstack != null && itemstack.getItem() instanceof ItemEditableBook)
-            {
+            if (itemstack != null && itemstack.getItem() instanceof ItemEditableBook) {
                 aitemstack[i] = itemstack;
                 break;
             }
@@ -124,4 +124,4 @@
 
         return aitemstack;
     }
-}
+}
